crates/core/tedge_mapper/src/c8y_converter.rs

```diff
@@ -129,6 +188,58 @@ impl Converter for CumulocityConverter {
             inventory_fragments_message,
         ])
     }
+
+    /// Detect and sync any alarms that were raised/cleared while this mapper process was not running.
+    /// Reconciliation is done by maintaining an internal journal of all the alarms processed by this mapper,
+    /// which is compared against all the live alarms seen by the mapper on every startup.
+    ///
+    /// All the live alarms are received from tedge/alarms topic on startup.
+    /// Similarly, all the previously processed alarms are received from c8y-internal/alarms topic.
+    /// Reconiciliation detects the difference between these two sets, which are the missed messages.
+    ///
+    /// An alarm that is present in c8y-internal/alarms, but not in tedge/alarms topic
+    /// is assumed to have been cleared while the mapper process was down.
+    /// Similarly, an alarm that is present in tedge/alarms, but not in c8y-internal/alarms topic
+    /// is one that was raised while the mapper process was down.
+    /// An alarm present in both, if their payload is the same, is one that was already processed before the restart
+    /// and hence can be ignored during reconcicilation.
+    fn sync_messages(&mut self) -> Vec<Message> {
```

Comment:
It couldn't be local variables in this function because it's the `convert` methods that populates those maps and this `sync` method just consumes these populated maps. So I needed these maps shared between these functions and that's why it was kept as fields of the `Converter`.

But you're right that we don't need to keep these maps beyond the sync method call and that's why I had used drain. In interpreted the "leaves the memory allocated for later reuse" as it just leaves the memory for an empty map allocated. Didn't expect them to leave the memory for **all entries** allocated for later reuse.

Reply:
That's fine, as an alternative it would be possible that one of these methods takes a generator function which takes a map of messages and produces vec of messages to sent.
I could provide example code later on, though.

I'm curious to see the enum impl.